lib/src/search.dart

```diff
@@ -0,0 +1,203 @@
+
+    allMatches.sort((a, b) {
+      // Exact match vs substring is king. If the user has typed the whole term
+      // they are searching for, but it isn't at the top, they cannot type any
+      // more to try and find it.
+      var comparison = a.matchPosition - b.matchPosition;
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer packages higher in the package order.
+      comparison = packageOrderPosition(a.item.packageName) -
+          packageOrderPosition(b.item.packageName);
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer top-level elements to library members to class (etc.) members.
+      comparison = a.item._scope - b.item._scope;
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer non-overrides to overrides.
+      comparison = a.item.overriddenDepth - b.item.overriddenDepth;
+      if (comparison != 0) {
+        return comparison;
+      }
+
+      // Prefer shorter names to longer ones.
+      return a.item.name.length - b.item.name.length;
+    });
+
+    return allMatches.map((match) => match.item).toList();
+  }
+}
+
+class IndexItem {
+  final String name;
+  final String qualifiedName;
+
+  // TODO(srawlins): Store the integer of in the package order instead of this
+  // String. The Strings bloat the `index.json` file and keeping duplicate
+  // parsed Strings in memory is expensive.
+  final String packageName;
+  final String type;
+  final String? href;
+  final int overriddenDepth;
+  final String? desc;
+  final EnclosedBy? enclosedBy;
+
+  IndexItem._({
+    required this.name,
+    required this.qualifiedName,
+    required this.packageName,
+    required this.type,
+    required this.desc,
+    required this.href,
+    required this.overriddenDepth,
+    required this.enclosedBy,
+  });
+
+  // Example Map structure:
+  //
+  // ```dart
+  // {
+  //   "name":"dartdoc",
+  //   "qualifiedName":"dartdoc",
+  //   "href":"dartdoc/dartdoc-library.html",
+  //   "type":"library",
+  //   "overriddenDepth":0,
+  //   "packageName":"dartdoc"
+  //   ["enclosedBy":{"name":"Accessor","type":"class"}]
+  // }
+  // ```
+  factory IndexItem.fromMap(Map<String, dynamic> data) {
+    // Note that this map also contains 'packageName', but we're not currently
+    // using that info.
+
+    EnclosedBy? enclosedBy;
+    if (data['enclosedBy'] != null) {
+      final map = data['enclosedBy'] as Map<String, Object>;
+      enclosedBy = EnclosedBy._(
+          name: map['name'] as String,
+          type: map['type'] as String,
+          href: map['href'] as String);
+    }
+
+    return IndexItem._(
+      name: data['name'],
+      qualifiedName: data['qualifiedName'],
+      packageName: data['packageName'],
+      href: data['href'],
+      type: data['type'],
+      overriddenDepth: (data['overriddenDepth'] as int?) ?? 0,
+      desc: data['desc'],
+      enclosedBy: enclosedBy,
+    );
+  }
+
+  int get _scope =>
```

Comment:
A comment explaining this would be nice.

Reply:
Good call, done! And added a TODO, as its not well-tested.